timemachine/cpp/src/local_md_utils.cu

```diff
@@ -0,0 +1,147 @@
+        fanned_potential != nullptr) {
+        return true;
+    } else if (std::shared_ptr<SummedPotential> summed_potential = std::dynamic_pointer_cast<SummedPotential>(pot);
+               summed_potential != nullptr) {
+        return true;
+    }
+    return false;
+}
+
+// Recursively populate nb_pots potentials with the NonbondedAllPairs
+void get_nonbonded_all_pair_potentials(
+    std::vector<std::shared_ptr<BoundPotential>> input, std::vector<std::shared_ptr<BoundPotential>> &nb_pots) {
+    for (auto pot : input) {
+        if (std::shared_ptr<FanoutSummedPotential> fanned_potential =
+                std::dynamic_pointer_cast<FanoutSummedPotential>(pot->potential);
+            fanned_potential != nullptr) {
+            std::vector<double> h_params(pot->size());
+            if (pot->size() > 0) {
+                pot->d_p->copy_to(&h_params[0]);
+            }
+            std::vector<int> shape{pot->size()};
+            std::vector<std::shared_ptr<BoundPotential>> flattened_bps;
+            for (auto summed_pot : fanned_potential->get_potentials()) {
+                if (is_summed_potential(summed_pot) || is_nonbonded_all_pairs_potential(summed_pot)) {
+                    flattened_bps.push_back(
+                        std::shared_ptr<BoundPotential>(new BoundPotential(summed_pot, shape, &h_params[0])));
+                }
+            }
+            get_nonbonded_all_pair_potentials(flattened_bps, nb_pots);
+            continue;
+        } else if (std::shared_ptr<SummedPotential> summed_potential =
+                       std::dynamic_pointer_cast<SummedPotential>(pot->potential);
+                   summed_potential != nullptr) {
+            std::vector<double> h_params(pot->size());
+            int i = 0;
+            int offset = 0;
+            if (pot->size() > 0) {
+                pot->d_p->copy_to(&h_params[0]);
+            }
+            std::vector<int> shape(1);
+
+            std::vector<std::shared_ptr<BoundPotential>> flattened_bps;
+            std::vector<int> param_sizes = summed_potential->get_parameter_sizes();
+            for (auto summed_pot : summed_potential->get_potentials()) {
+
+                if (is_summed_potential(summed_pot) || is_nonbonded_all_pairs_potential(summed_pot)) {
+                    std::vector<double> slice(h_params.begin() + offset, h_params.begin() + offset + param_sizes[i]);
+                    shape[0] = param_sizes[i];
+                    flattened_bps.push_back(
+                        std::shared_ptr<BoundPotential>(new BoundPotential(summed_pot, shape, &slice[0])));
+                }
+
+                offset += param_sizes[i];
+                i++;
+            }
+            get_nonbonded_all_pair_potentials(flattened_bps, nb_pots);
+            continue;
+        }
+        if (is_nonbonded_all_pairs_potential(pot->potential)) {
+            nb_pots.push_back(pot);
+        }
+    }
+}
```

Comment:
I can't review this confidently, but it looks reasonable.

Reply:
If you want to walk through this together offline let me know.